pkg/addon/registry.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package addon
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	velatypes "github.com/oam-dev/kubevela/apis/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var registryConfigMapName = "vela-addon-registry"
+var registriesKey = "registries"
+
+// Registry represent a addon registry model
+type Registry struct {
+	Name string `json:"name"`
+
+	Git *GitAddonSource `json:"git,omitempty"`
+	Oss *OSSAddonSource `json:"oss,omitempty"`
+}
+
+// RegistryDataStore CRUD addon registry data in configmap
+type RegistryDataStore interface {
+	ListRegistries(context.Context) ([]Registry, error)
+	AddRegistry(context.Context, Registry) error
+	DeleteRegistry(context.Context, string) error
+	UpdateRegistry(context.Context, Registry) error
+	GetRegistry(context.Context, string) (Registry, error)
+}
+
+// NewRegistryDataStore get RegistryDataStore operation interface
+func NewRegistryDataStore(cli client.Client) RegistryDataStore {
+	return registryImpl{cli}
+}
+
+type registryImpl struct {
+	client client.Client
+}
+
+func (r registryImpl) ListRegistries(ctx context.Context) ([]Registry, error) {
+	cm := &v1.ConfigMap{}
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: velatypes.DefaultKubeVelaNS, Name: registryConfigMapName}, cm); err != nil {
+		return nil, err
+	}
+	registries := map[string]Registry{}
+	if err := json.Unmarshal([]byte(cm.Data[registriesKey]), &registries); err != nil {
```

Comment:
It would be better if we check whether `registriesKey` exists in the data of ConfigMap as we can throw clear error messages like `missing {registriesKey} key`. If we could hit error message like `unexpected end of JSON input`.

Reply:
Nice catch, fixed